### PR TITLE
Check HTTP status code, document 0 value, casts

### PR DESCRIPTION
**Pull request by gperciva**

### CODE REVIEW

File: lib/proto_s3/proto_s3_client.c [hunk 1]
```diff
@@ -481,9 +482,12 @@ callback_head(void * cookie, uint8_t * buf, size_t buflen)
 			lens = len;
 	}
 
+	/* Sanity check. */
+	assert((status == 0) || ((status >= 100) && (status <= 599)));
```
Comment (cperciva) [inline lib/proto_s3/proto_s3_client.c]: Absolutely not.  You should never `assert` what you received over the wire -- that would make it possible for the server to crash us.

This should be checked the same way as we check if a response is valid: `if (...) BAD("HEAD", "invalid HTTP status")`.

**Comment by gperciva** [comment]
Revised PR, rewriting history.  I added a separate commit for checking the `status` from packets in `proto_s3_client.c`.

File: lib/proto_s3/proto_s3_client.c [hunk 2]
```diff
@@ -127,6 +127,10 @@ callback_put(void * cookie, uint8_t * buf, size_t buflen)
 		/* Parse the packet. */
 		status = be32dec(&buf[0]);
 
+		/* Check status. */
+		if ((status < 100) || (status > 599))
```
Comment (cperciva) [inline lib/proto_s3/proto_s3_client.c]: Can't we have `status == 0` here?  (And below)

**Comment by cperciva** [comment]
LGTM, please rebase.

**Comment by gperciva** [comment]
This PR has been rebased and is ready for merge.

**Pull request merged**